Mathlib/LinearAlgebra/Matrix/PosDef.lean

```diff
@@ -38,59 +43,220 @@ open scoped Matrix
     (M.submatrix e e).PosSemidef := by
-  refine' ⟨hM.1.submatrix e, fun x => _⟩
-  have : (M.submatrix (⇑e) e).mulVec x = (M.mulVec fun i : n => x (e.symm i)) ∘ e := by
-    ext i
-    dsimp only [(· ∘ ·), mulVec, dotProduct]
-    rw [Finset.sum_bij' (fun i _ => e i) _ _ fun i _ => e.symm i] <;>
-      simp only [eq_self_iff_true, imp_true_iff, Equiv.symm_apply_apply, Finset.mem_univ,
-        submatrix_apply, Equiv.apply_symm_apply]
-  rw [this]
-  convert hM.2 fun i => x (e.symm i) using 3
-  unfold dotProduct
-  rw [Finset.sum_bij' (fun i _ => e i) _ _ fun i _ => e.symm i] <;>
-  simp
+  classical
+  rw [(by simp : M = 1 * M * 1), submatrix_mul (e₂ := id) (he₂ := Function.bijective_id),
+    submatrix_mul (e₂ := id) (he₂ := Function.bijective_id), submatrix_id_id]
+  simpa only [conjTranspose_submatrix, conjTranspose_one] using
+    conjTranspose_mul_mul_same hM (Matrix.submatrix 1 id e)
 #align matrix.pos_semidef.submatrix Matrix.PosSemidef.submatrix
 
+/-- The eigenvalues of a positive semi-definite matrix are non-negative -/
+lemma eigenvalues_nonneg [DecidableEq n] {A : Matrix n n 𝕜}
+    (hA : Matrix.PosSemidef A) (i : n) : 0 ≤ hA.1.eigenvalues i :=
+  (hA.re_dotProduct_nonneg _).trans_eq (hA.1.eigenvalues_eq _).symm
+
+theorem transpose {M : Matrix n n R} (hM : M.PosSemidef) : Mᵀ.PosSemidef := by
+  refine ⟨IsHermitian.transpose hM.1, fun x => ?_⟩
+  convert hM.2 (star x) using 1
+  rw [mulVec_transpose, Matrix.dotProduct_mulVec, star_star, dotProduct_comm]
+
+section sqrt
+
+variable [DecidableEq n] {A : Matrix n n 𝕜} (hA : PosSemidef A)
+
+/-- The positive semidefinite square root of a positive semidefinite matrix -/
+@[pp_dot]
+noncomputable def sqrt : Matrix n n 𝕜 :=
+  hA.1.eigenvectorMatrix * diagonal ((↑) ∘ Real.sqrt ∘ hA.1.eigenvalues) * hA.1.eigenvectorMatrixᴴ
+
+open Lean PrettyPrinter.Delaborator SubExpr in
+/-- Custom elaborator to produce output like `(_ : PosSemidef A).sqrt` in the goal view. -/
+@[delab app.Matrix.PosSemidef.sqrt]
+def delabSqrt : Delab :=
+  whenPPOption getPPNotation <|
+  whenNotPPOption getPPAnalysisSkip <|
+  withOptionAtCurrPos `pp.analysis.skip true do
+    let e ← getExpr
+    guard <| e.isAppOfArity ``Matrix.PosSemidef.sqrt 7
+    let optionsPerPos ← withNaryArg 6 do
+      return (← read).optionsPerPos.setBool (← getPos) `pp.proofs.withType true
+    withTheReader Context ({· with optionsPerPos}) delab
+
+-- test for custom elaborator
+variable [DecidableEq n] {A : Matrix n n 𝕜}  (hA : PosSemidef A)
+/--
+info: (_ : PosSemidef A).sqrt : Matrix n n 𝕜
+-/
+#guard_msgs in
+#check (id hA).sqrt
+
+lemma posSemidef_sqrt : PosSemidef hA.sqrt := by
+  apply PosSemidef.mul_mul_conjTranspose_same
+  refine posSemidef_diagonal_iff_nonneg.mpr fun i ↦ ?_
+  rw [Function.comp_apply, IsROrC.nonneg_iff]
+  constructor
+  · simp only [IsROrC.ofReal_re]
+    exact Real.sqrt_nonneg _
+  · simp only [IsROrC.ofReal_im]
+
+lemma sq_sqrt : hA.sqrt ^ 2 = A := by
```

Comment:
This should probably be `simp`, along with `sqrt_mul_self` (matching `Real.sqrt_mul_self`) for good measure

Reply:
Done